first draft  of custom deployment instructions (#3893)
pachyderm/pachyderm: doc/deployment/deploy_custom.md

> # Custom Deployments

Reviewer: Hi level comment:

All of the information you have here is good, but I think we can improve the information architecture a bit. It feels like you lost the forest for the trees given that this has been an ongoing side project for so long. The 2nd half of the doc is excellent, but the first half isn't clear where it fits in the overall process and other deploy docs:

E.g. 
1. On_premises doc
2. Custom deploy doc (this one)
3. ? --  Is there a 3rd?

We probably just need to put a more organized opening, TOC, etc.

Author: We chatted about this on the phone, and I also discussed with @svekars.  Every doc needs to be somewhat self-contained, with preambles that bring the reader up to speed.

> 

Author: Noted.  I've also been using semantic line breaks when writing.  Is that something you want to adopt?

[Google Cloud Platform (GCP)](https://pachyderm.readthedocs.io/en/latest/deployment/google_cloud_platform.html), or 
[Microsoft Azure](https://pachyderm.readthedocs.io/en/latest/deployment/azure.html), 
Pachyderm's `pachctl deploy` includes commands for those providers (`amazon`, `google`, and `microsoft`, respectively).
You should use those commands when deploying to those cloud infrastructures.
You can customize cloud provider deployments extensively through flags available for each provider.

Pachyderm also includes `pachctl deploy custom` for creating customized deployments for cloud providers or on-premises use.
It's usually invoked with the `--dry-run` flag, 
with the command's standard output directed into a series of customization scripts or a file for editing.

This document covers using `pachctl deploy custom ... --dry-run` to create a manifest for a custom, on-premises deployment.
We won't directly address automating your deployment by means of customization scripts,
but we do encourage you to treat your [infrastructure as code](./on-premises.html#infrastructure-as-code).

> 

Reviewer: Before starting to break down the manifest that is generated, we probably need to break down the required and optional flags in deploy custom. You cover some of them in later sections in terms of how they populate the manifest, but we should probably upfront tell a user how to run the command. 

Also, there should likely be an example command and corresponding manifest (or at last snippets of the manifest).

Author: You're absolutely right.  I need to first narrow the choices to the set of deployment commands likely to be used on-premises.  I don't think that was "google", as the old doc had, but more likely AWS arguments.

When planning this documentation, we wanted the followon to be a set of PR's revamping the user experience for on-premises deployments. I'm coming to the conclusion that the user experiences for cloud vs on-premises deployments are substantially different.

## Anatomy of a Pachyderm deployment manifest

When you run the `pachctl deploy ...` command with the `--dry-run` flag,
you're generating a JSON-encoded Kubernetes manifest in one stream to standard output. 
That manifest itself consists of a number of smaller manifests,
corresponding to a particular aspect of a Pachyderm deployment.

Pachyderm has three sets of application components usually deployed:
- `pachd`, the main Pachyderm pod
- `etcd`, the administrative datastore for `pachd`
- `dash`, the web-based enterprise ui for Pachyderm
We'll explore all of these within this document.

> In general, there are two categories of manifests in the file,

Author: I"d like to shy away from inventing the term "submanifest" here, which I can't find anywhere in the kubernetes literature.  Unless you know better?

> In general, there are two categories of manifests in the file,

Reviewer 2: I've seen people using the term "submanifest" when they describe exactly what you are trying to describe here but I agree it's not widely used. Maybe: "A typical Pachyderm manifest file includes the following types of manifests:"

roles-and-permissions-related and application-related.

> ## Roles-and-permissions-related manifests

Author: I'm old-fashioned: I hyphenate adjectival phrases.  I'll bow to your editorial policy. :-)

### ServiceAccount

> Usually at the top of the file, this manifest has the `kind` key set to `ServiceAccount`. 

Author: Turning "at the top of the file" into an appositive changes the meaning of this sentence.  It's the manifest itself that's typically at the top of the file, not the kind key that's set to "ServiceAccount".

[ServiceAccounts](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/) are a way Kubernetes can assign namespace-specific privileges to applications in a lightweight way.

> Pachyderm's service account is called, appropriately enough, `pachyderm`.

Author: Oh, our editorial voice is having a whimsy-ectomy.  But, that's OK. :-)

### Role or ClusterRole

> The next manifest will usually be of `kind` `Role` or `ClusterRole`, 

Author: "Role" or "ClusterRole" describes the "kind" of the manifest it is; it wasn't a list of "kind", "Role" and "ClusterRole"...

> depending on whether you  used the `--local-roles` flag `pachctl deploy` command.

Reviewer: sentence structure is weird here. Just remove "`pachctl deploy` command"

Reviewer: Add: At a high level, what does "role" section of manifest define? 1 sentence is plenty.

Author: I believe this is missing a "to" before "`pachctl deploy` command", which would make it clearer, but I'll rephrase into shorter, clearer sentences.

### RoleBinding or ClusterRoleBinding

> This manifest binds the `Rule` or `ClusterRole` to the `ServiceAccount` created above.

Reviewer: typo: rule -->role

Author: Fixed below, in a set of @svekars comments.

> This manifest binds the `Rule` or `ClusterRole` to the `ServiceAccount` created above.

Reviewer 2: ```suggestion
```

Author: Why delete this line?  I think it needs rephrasing, but the function of the manifest should be explained.

Reviewer 2: I think I might have edited this in a separate comment but forgot to add. LGTM!

## Application-related

### PersistentVolume

> If you don't use [StatefulSets](./on_premises.html#statefulsets) to deploy Pachyderm

Reviewer: This whole section is one sentence. Break it down and make it clearer.
e.g.:
Pachyderm's etcd needs a volume to stare data. You can either specify the `--dynamic-etcd-nodes` flag ... or the `--persistent-disk` flag. 

If you do A, then...

If you do B, then...

> If you don't use [StatefulSets](./on_premises.html#statefulsets) to deploy Pachyderm

Reviewer 2: ```suggestion
If you did not use [StatefulSets](./on_premises.html#statefulsets) to deploy Pachyderm,
```

> (that is, you don't specify `--dynamic-etcd-nodes` flag), 

Reviewer 2: ```suggestion
that is, you do not specify `--dynamic-etcd-nodes` flag, 
```

> the value you specify for `--persistent-disk` will cause `pachctl` to write a manifest for creating a [`PersistentVolume`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) for Pachyderm's `etcd` to use in its [`PersistentVolumeClaim`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims).

Reviewer 2: ```suggestion
the value that you specify for `--persistent-disk` causes `pachctl` to write a manifest for creating a [`PersistentVolume`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) that Pachyderm's `etcd` uses in its [`PersistentVolumeClaim`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims).
```

### PersistentVolumeClaim

> This is the `PersistentVolumeClaim` that Pachyderm's `etcd` will use if you don't deploy using [StatefulSets](./on_premises.html#statefulsets).

Reviewer: It's weird to have this section and the above framed is the form of a negative.. "you'll see this if you don't do...". Is there a reasonably way to do it differently?

Author: My intent here is to establish StatefulSets as the norm.  I'll rephrase it accordingly.

> This is the `PersistentVolumeClaim` that Pachyderm's `etcd` will use if you don't deploy using [StatefulSets](./on_premises.html#statefulsets).

Reviewer 2: ```suggestion
Pachyderm's `etcd` uses this `PersistentVolumeClaim` unless you deploy a [StatefulSets](./on_premises.html#statefulsets).
```

Author: "deploy a StatefulSets" sounded clumsy, so I reverted the phrasing to "deploy using StatefulSets".

> You'll see this manifest's name in the `etcd` Deployment manifest, below.

Reviewer 2: ```suggestion
See this manifest's name in the `etcd` Deployment manifest, below.
```

### StorageClass

> If you *do* use [StatefulSets](./on_premises.html#statefulsets) to deploy Pachyderm

Reviewer 2: ```suggestion
If you *do* use [StatefulSets](./on_premises.html#statefulsets) to deploy Pachyderm,
```

Author: @svekars: I'm worried that relative links won't work properly on github.  Do we have a way to make this work without resorting to full urls?

Reviewer 2: I've seen this issue in the past and I don't know of a good way to fix it. If you have a relative link to an html file, the link won't work on GitHub. Replacing these links with full links is not a solution either. You could try to add a footnote and give the correct relative link to the .md file at the bottom of the page, but it's an ugly solution, IMHO.

Reviewer 2: We can add a disclaimer in the docs/index.md file that says something like: "If you read this documentation on GitHub, some links might not work. For complete documentation, see https://pachyderm.readthedocs.io."

> (that is, you use `--dynamic-etcd-nodes` flag), 

Reviewer 2: ```suggestion
that is, you use `--dynamic-etcd-nodes` flag, 
```

> this manifest, if needed, will specify the kind of storage and provisioner that's appropriate for what you've specified in the `--persistent-disk` flag. 

Reviewer: why is "if needed" there?

Author: When deploying to Azure, we don't generate this manifest. The provisioner appears to be hardwired in some way, so it doesn't need to be specified. It's unclear to me if other versions of Kubernetes could have that same behavior.

> this manifest, if needed, will specify the kind of storage and provisioner that's appropriate for what you've specified in the `--persistent-disk` flag. 

Reviewer 2: ```suggestion
this manifest specifies the kind of storage and provisioner that is appropriate for what you have specified in the `--persistent-disk` flag. 
```

Author: I've rephrased this to make it clear when you won't see this. (See the comment above in reply to @JoeyZwicker)

### Service

> You'll see three [`Service`](https://kubernetes.io/docs/concepts/services-networking/service/) manifests in a Pachyderm deployment. 

Reviewer 2: ```suggestion
In a typical Pachyderm deployment, you see three [`Service`](https://kubernetes.io/docs/concepts/services-networking/service/) manifests. 
```

Services are how Kubernetes exposes Pods to the network.

> If you  use [StatefulSets](./on_premises.html#statefulsets) to deploy Pachyderm

Reviewer 2: ```suggestion
If you  use StatefulSets to deploy Pachyderm,
```

Author: Why remove the link?

> (that is, you use `--dynamic-etcd-nodes` flag), 

Reviewer 2: ```suggestion
that is, you use `--dynamic-etcd-nodes` flag,
```

> you'll see one for `etcd-headless`, one for `pachd`, and one for `dash`.

Reviewer 2: ```suggestion
Pachyderm deploys one Service for `etcd-headless`, one for `pachd`, and one for `dash`.
```

> A static deployment will have `Services` for `etcd`, `pachd`, and `dash`.

Reviewer 2: ```suggestion
A static deployment has `Services` for `etcd`, `pachd`, and `dash`.
```

> The `dash` `Service` and `Deployment` will be omitted if the `--no-dashboard` is used.

Reviewer 2: ```suggestion
If you use the `--no-dashboard` flag, Pachyderm does not create the `dash` `Service` and `Deployment`.
```

Likewise, if `--dashboard-only` is specified,

> manifests for the Pachyderm enterprise UI only will be generated. 

Reviewer 2: ```suggestion
Pachyderm generates the manifests for the Pachyderm enterprise UI only. 
```

> The most common items to edit in `Service` manifests are the `NodePort` values for various services, 

Reviewer 2: ```suggestion
The most common items that you can edit in `Service` manifests are the `NodePort` values for various services, 
```

and the `containerPort` values for `Deployment` manifests.

> It will be necessary to add environment variables to a `Deployment` or `StatefulSet` object to make your `containerPort` values work properly.

Reviewer 2: ```suggestion
To make your `containerPort` values work properly, add environment variables to a `Deployment` or `StatefulSet` object.
```

> A good example to check against is [OpenShift](./openshift.html).

Reviewer 2: ```suggestion
You can verify this functionality in the [OpenShift](./openshift.html) example.
```

### The Pachyderm pods

#### Deployment 

> A [`Deployment`](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) is how it's declared to Kubernetes what the final state of an application's pods should be.

Reviewer 2: ```suggestion
A [`Deployment`](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) declares the desired state of application pods to Kubernetes.
```

> For a static deployment, 

Reviewer 2: ```suggestion
If you configure a static deployment,
```

> there will three [`Deployment`](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) manifests, for `etcd`, `pachd`, and `dash`.

Reviewer 2: ```suggestion
Pahyderm deploys `Deployment` manifests for `etcd`, `pachd`, and `dash`.
```

> If you specify `--dynamic-etd-nodes` when doing your deploy, you'll only see two `Deployment`s, 

Reviewer 2: ```suggestion
If you specify `--dynamic-etd-nodes`, Pachyderm deploys the `pachd` and `dash` as `Deployment`
```

> `pachd` and `dash`,

Reviewer 2: ```suggestion
```

> and  `etcd` will be deployed using a `StatefulSet`.

Reviewer 2: ```suggestion
and `etcd` as a`StatefulSet`.
```

> The `dash` `Service` and `Deployment` will be omitted if the `--no-dashboard` is used.

Reviewer 2: ```suggestion
If you run the deploy command with the `--no-dashboard` flag, Pachyderm omits the deployment of the `dash` `Service` and `Deployment`.
```

#### StatefulSet

> For a `--dynamic-etcd-nodes` deployment, the `etcd` `Deployment` manifest will be replaced with a `StatefulSet`.

Reviewer 2: ```suggestion
For a `--dynamic-etcd-nodes` deployment, Pachyderm replaces the `etcd` `Deployment` manifest with a `StatefulSet`.
```

### Secret

The final manifest is a Kubernetes [`Secret`](https://kubernetes.io/docs/concepts/configuration/secret/).

> It's used by Pachyderm to store the credentials necessary access object storage.

Reviewer 2: ```suggestion
Pachyderm uses the `Secret` to store the credentials that are necessary to access object storage.
```

> It uses the command-line arguments to the `pachctl deploy` command to store the parameters, like region, secret, token and endpoint, used to access an object store. 

Reviewer 2: ```suggestion
The final manifest uses the command-line arguments that you submit to the `pachctl deploy` command to store the parameters, like region, secret, token, and endpoint that are used to access an object store. 
```

The exact values in the secret depend on the kind of object store you configure for your deployment.

> You can update the values after deployment either using `kubectl` to deploy a new `Secret`

Reviewer 2: ```suggestion
You can update the values after the deployment either by using `kubectl` to deploy a new `Secret`
```

### Preparing your environment

See the [introduction to on-premises deployment](./on_premises.html) for steps that you need to take 
before creating a custom Pachyderm deployment manifest. 
That section also explains differences among static persistent volumes, 
StatefulSets and StatefulSets with StorageClasses, 
as well as the meanings of the variables, 
such as  `PVC_STORAGE_SIZE` and `OS_ENDPOINT`, 
used in examples below.

## Creating a Pachyderm deployment manifest

Use the `pachctl deploy custom` command to create a custom manifest.custom`.

**Example**
```
pachctl deploy custom --persistent-disk <persistent disk backend> --object-store <object store backend> \
    <persistent disk arg1>  <persistent disk arg 2> \
    <object store arg 1> <object store arg 2>  <object store arg 3>  <object store arg 4> \
    [[--dynamic-etcd-nodes n] | [--static-etcd-volume <volume name>]]
    [optional flags]
```

> 

Reviewer: There should be more of an opening to set the stage here. Something along the lines of:

"The `pachctl deploy custom` command takes a bunch of different flags in order to generate an appropriate manifest for your infrastructure. The flags broadly fall into a few categories:

- `--persistent-disk`: configure the storage resource that is used by etcd. Pachyderm uses etcd to manage administrative metadata. User data is stored in an object store, not etcd.
- `--object-store`: configure the object store that Pachyderm uses for storing all user data that you want versioned and managed.
- `optional flags`: Other configurations and details that can be customized."

These 3 bullets should be the headers for your other sections. They basically already are, so thats good, but this one of those "tell them what you're going to tell them" pieces that some of your big docs are initially missing. In this case you put a TOC to try to offer this, but as svetlana pointed out, it was way too big and confusing when you can just do in a simple intro section as above.

- _bucket-name_: The name of the bucket, without the `s3://` prefix or a trailing forward slash (`/`).
- _access-key_: The user access ID that is used to access the object store.
- _secret-key_: The associated password that is used with the user access id to access the object store.
- _endpoint_: The hostname and port that are used to access the object store, in <hostname>:<port> format.

#### Example invocation with pv and object store
The example on-premises cluster will use an on-premises MinIO object store with the following configuration parameters:
  - SSL turned on,
  - S3v4 signatures,
  - the endpoint `minio:9000`
  - the access key `OBSIJRBE0PP2NO4QOA27`,
  - the secret key `tfteSlswRu7BJ86wekitnifILbZam1KYY3TG`,
  - and a bucket named `pachyderm-bucket`.
The deployment command uses the following flags:
```
pachctl deploy custom --persistent-disk aws --object-store s3 \
    foobar 10 \
    pachyderm-bucket  'OBSIJRBE0PP2NO4QOA27' 'tfteSlswRu7BJ86wekitnifILbZam1KYY3TG' 'minio:9000' \
    --dynamic-etcd-nodes 10
    [optional flags]
```
In the example command above, some of the arguments may contain characters that the shell could interpret in single-quotes.

> ### Additional flags

Reviewer 2: all the headings under this *Additional flags* can be put in a table or a list. We also need an intro text before the table. Something like:
```
This table lists the <blah> flags that you can use to customize your Pachyderm deployment.
```

> ### Additional flags

Reviewer 2: Also, the title could be less generic. Maybe *System-wide Flags*?

```
pachctl deploy custom --persistent-disk aws --object-store s3 \
    foobar 10 \
    pachyderm-bucket  'OBSIJRBE0PP2NO4QOA27' 'tfteSlswRu7BJ86wekitnifILbZam1KYY3TG' 'minio:9000' \
    --dynamic-etcd-nodes 10
    --local-roles --output yaml  --dry-run > custom_deploy.yaml
```

For more information about the contents of the `custom_deploy.yaml` file,
see [Anatomy of a Pachyderm deployment manifest](#anatomy-of-a-pachyderm-deployment-manifest).

## Anatomy of a Pachyderm deployment manifest

By default, when you run the `pachctl deploy ...` command with the `--dry-run` flag,
you are generating a JSON-encoded Kubernetes manifest in one stream to standard output. 
That manifest consists of smaller manifests,
each of which corresponds to a particular aspect of a Pachyderm deployment.

Pachyderm deploys the following sets of application components:
- `pachd`, the main Pachyderm pod
- `etcd`, the administrative datastore for `pachd`
- `dash`, the web-based enterprise ui for Pachyderm Enterprise Edition

In general, there are two categories of manifests in the file,
roles-and-permissions-related and application-related. 

## Roles and permissions manifests

> 

Reviewer 2: all the headings under this *Roles and permissions manifests* can be put in a table. We also need an intro text before the table. Something like:
```
This table lists all the roles and permissions sections that Pachyderm configures in the deployment manifest.
```

This section walks you through the roles and permissions manifests.

### ServiceAccount

Typically at the top of the file, a roles and permissions manifest has the `kind` key set to `ServiceAccount`. 
[ServiceAccounts](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/) are a way Kubernetes can assign namespace-specific privileges to applications in a lightweight way.
The Pachyderm's service account is called  `pachyderm`.

### Role or ClusterRole

Depending on whether you used the `--local-roles` flag or not, the next manifest `kind` is `Role` or `ClusterRole`.

### RoleBinding or ClusterRoleBinding

This manifest binds the `Role` or `ClusterRole` to the `ServiceAccount` created above.

> ## Application-related

Reviewer 2: all the headings under this *Application-related* can be put in a table. We also need an intro text before the table. Something like:
```
This table lists all the application-related sections that Pachyderm configures in the deployment manifest.
```

that you should use for `etcd` in your Kubernetes deployment.
If you need to use a different one than the default setting, 
you can use the `--etcd-storage-class` flag to specify the StorageClass.

`--static-etc-volume` is used when your Kubernetes installation has not been configured to use StatefulSets.
When you specify `--static-etc-volume` flag, Pachyderm creates a static volume for `etcd`. 
Pachyderm creates a PV with a spec appropriate for each of the cloud providers:

- `aws`: awsElasticBlockStore for Amazon Web Services
- `azure`: azureDisk for Microsoft Azure
- `google`: gcePersistentDisk for Google Cloud Storage

These choices are not applicable to  most on-premises deployments.
To determine the correct choices for your on-prem infrastructure, 
consult with your Kubernetes administrators
and edit your manifest manually.

For more information, see [storage manifests](#persistentvolume).

#### Persistent disk parameters

Regardless whether you choose to deploy with StatefulSets or static volumes,
the `--persistent-disk` flag takes two arguments
that you specify right after the single argument to the `--object-store` flag.

> Although the first argument is required, Pachyderm ignores it.

Reviewer: Later one you say:

> Note: we use any-string as the first argument above because, while the deploy custom command expects 6 arguments, it will ignore the first argument when deploying with StatefulSets.

If this is the technical requirement for why this exists, i can be make much clearer as you do in the note later.

These arguments must be placed immediately after [the persistent disk configuration parameters](#persistent-disk-parameters).

- _bucket-name_: The name of the bucket, without the `s3://` prefix or a trailing forward slash (`/`).
- _access-key_: The user access ID that is used to access the object store.
- _secret-key_: The associated password that is used with the user access id to access the object store.
- _endpoint_: The hostname and port that are used to access the object store, in <hostname>:<port> format.

#### Example invocation with pv and object store
The example on-premises cluster will use an on-premises MinIO object store with the following configuration parameters:
  - SSL turned on,
  - S3v4 signatures,
  - the endpoint `minio:9000`
  - the access key `OBSIJRBE0PP2NO4QOA27`,
  - the secret key `tfteSlswRu7BJ86wekitnifILbZam1KYY3TG`,
  - and a bucket named `pachyderm-bucket`.
The deployment command uses the following flags:
```
pachctl deploy custom --persistent-disk aws --object-store s3 \
    foobar 10 \
    pachyderm-bucket  'OBSIJRBE0PP2NO4QOA27' 'tfteSlswRu7BJ86wekitnifILbZam1KYY3TG' 'minio:9000' \
    --dynamic-etcd-nodes 10
    [optional flags]
```

> In the example command above, some of the arguments may contain characters that the shell could interpret in single-quotes.

Reviewer: "could interpret in single quotes"? What does this mean. Does it mean I need single quotes, double quotes, on whic arguments?

### Additional flags

This section describes all the additional flags that you can use to deploy a custom Pachyderm cluster.

#### Local vs cluster roles

You can use the `--local-roles` flag to change the kind of role the `pachyderm` service account uses 
from cluster-wide (`ClusterRole`) to namespace-specific (`Role`). 
Using `--local-roles` inhibits your ability to use the [coefficient parallelism](http://docs.pachyderm.io/en/latest/reference/pipeline_spec.html#parallelism-spec-optional) feature.
After you set the `--local-roles` flag, 
you might see the following message in the `pachd` pod Kubernetes logs:

```
ERROR unable to access kubernetes nodeslist, Pachyderm will continue to work but it will not be possible to use COEFFICIENT parallelism. error: nodes is forbidden: User "system:serviceaccount:pachyderm-test-1:pachyderm" cannot list nodes at the cluster scope
```

#### Resource requests and limits

Larger deployments might require you to configure more resources for `pachd` and `etcd`
or set higher limits for transient workloads.
The following flags set attributes which are passed on to Kubernetes directly through the produced manifest.

> Do not modify the default values of these flags for production deployments without consulting with Pachyderm support.

Reviewer: I dont think this is true. changing the resources for pachd and etcd doesn't require our help. If needed, we can recommend that they dont make them smaller, but users are capable of making them bigger on their own if needed, hopefully.

- `--etcd-cpu-request`: The number of CPU cores that Kubernetes allocates to `etcd`. Fractions are allowed. 
- `--etcd-memory-request`: The amount of memory that Kubernetes allocates to `etcd`. The SI suffixes are accepted as possible values.
- `--no-guaranteed`: Turn off QoS for `etcd` and `pachd`. Do not use this flag in production environments.
- `--pachd-cpu-request`: The number of CPU cores that Kubernetes allocates to `pachd`. Fractions are allowed. 
- `--pachd-memory-request`: The amount of memory that Kubernetes allocates to `pachd`. This flag accepts the SI suffixes.

> - `shards`: The maximum number of `pachd` nodes allowed in the cluster. Increasing this number from the default value of 16 might result in degraded performance.

Reviewer: Can you explain what this actually does to me? Is it literally just a weird limit on the total number of pachd nodes? that doesn't really make sense.

#### Enterprise Edition features

- `--dash-image`: The Docker image for the Pachyderm Enterprise Edition dashboard.
- `--image-pull-secret`: The name of a Kubernetes secret that Pachyderm uses to pull from a private Docker registry.
- `--no-dashboard`: Skip the creation of a manifest for the Enterprise Edition dashboard.
- `--registry`: The registry for Docker images.
- `--tls`:  A string in the `"<cert path>,<key path>"` format with the signed TLS certificate that is used for encrypting `pachd` communications.

#### Output formats

- `--dry-run`: Create a manifest and send it to standard output, but do not deploy to Kubernetes.
- `-o` or `--output`: An output format. You can choose from JSON (default) or YAML.

#### Logging
- `log-level`: The `pachd` verbosity level, from most verbose to least. You can set this parameter to `debug`, `info`, or `error`.
- `-v` or `--verbose`: Controls the verbosity of the `pachctl` invocation.

#### Complete example invocation

> Since we're deploying to a code infrastructure that works with YAML files,

Reviewer: this whole first paragraph is awkwardly worded. I get that we're just trying to show an end-to-end example. So just say so.  Also, the phrase "code infrastructure" get used here and elsewhere and it sounds awkward. I understand the tagline "infra as code", but its not a noun like you use keep using here (and elsewhere, please fix those too). 

"Here's a complete deploy command example of a custom deployment. In this example, the infrastructure expects a YAML manifest and uses `local-roles` instead of cluster-wide roles so there are optional flag sets for those constraints. " <-- Am i missing something or is this a much simpler phrasing to get the same point across?

```
pachctl deploy custom ..... --dry-run
```

    foobar 10 \
    pachyderm-bucket  'OBSIJRBE0PP2NO4QOA27' 'tfteSlswRu7BJ86wekitnifILbZam1KYY3TG' 'minio:9000' \
    --dynamic-etcd-nodes 10
    --local-roles --output yaml  --dry-run > custom_deploy.yaml
```

What does a file like `custom_deploy.yaml` have inside of it?
That's in the next section: 
a general exploration of all the deployment manifests that `pachctl deploy custom` produces.

## Anatomy of a Pachyderm deployment manifest

When you run the `pachctl deploy ...` command with the `--dry-run` flag,
you are generating, by default, a JSON-encoded Kubernetes manifest in one stream to standard output. 
That manifest consists of a number of smaller manifests,
each of which corresponds to a particular aspect of a Pachyderm deployment.

Pachyderm deploys the following sets of application components:
- `pachd`, the main Pachyderm pod
- `etcd`, the administrative datastore for `pachd`
- `dash`, the web-based enterprise ui for Pachyderm Enterprise Edition

In general, there are two categories of manifests in the file,
roles-and-permissions-related and application-related. 

## Roles and permissions manifests

### ServiceAccount

> Typically at the top of the file, a roles and permissions manifest has the `kind` key set to `ServiceAccount`. 

Author: "at the top of the file" isn't in apposition to "Typically", it's part of the adverbial phrase itself, as in "this section is typically at the top of the file".

### Preparing your environment

> Please see the [introduction to on-premises deployment](./on_premises.html) for steps you need to take prior to creating a custom Pachyderm deployment manifest.

Reviewer 2: ```suggestion
See the [introduction to on-premises deployment](./on_premises.html) for steps that you need to take before to creating a custom Pachyderm deployment manifest.
```

> ### Customizing through pachctl flags

Reviewer 2: ```suggestion
### Customizing `pachctl` flags
```

`pachctl` includes flags for customizing aspects of your deployment,
from memory and cpu requests for `etcd` and `pachd` to specifying a Kubernetes namespace.

> Please check what flags are available in your version of Pachyderm by running `pachctl deploy custom --help`.

Reviewer: I think this doc should include this information. maybe not every option, but at the very least the ones that are common or most confusing. This is less necessary if we update the --help to be better, but that'd need to be a separate PR.

Author: That was in our original plan for these docs: creating a set of PR's to reorganize and clarify those flags. 

I think the most common ones are the cpu and memory requests.  It would be nice to get those instrumented so we can collect data on how people use them...

> Please check what flags are available in your version of Pachyderm by running `pachctl deploy custom --help`.

Reviewer 2: ```suggestion
You can learn what flags are available in your version of Pachyderm by running `pachctl deploy custom --help`.
```

Flag that should be used with caution are marked as such.

> Please consult with your Kubernetes administrator and your Pachyderm support team when unsure of the effect of a flag.

Reviewer 2: ```suggestion
When you are unsure of the effect of a flag, consult with your Kubernetes administrator and your Pachyderm support team.
```

> Some flags, like `--image-pull-secret`, will require the creation and loading of Kubernetes manifests outside of `pachctl`.

Reviewer 2: ```suggestion
Some flags, such as `--image-pull-secret`, require the creation and loading of Kubernetes manifests outside of `pachctl`.
```

> ### Editing your manifest to customize it further

Author: The process of applying flags to the command line customizes the manifest, so you're in the process of customizing it when you run `pachctl deploy custom` with flags.  Editing it continues that process, so you're taking it further.

> This should only be approached if you are an experienced Kubernetes administrator.

Reviewer 2: ```suggestion
This functionality requires an experienced Kubernetes administrator.
```

If you are attempting a highly customized deployment, 

> please engage with your Pachyderm support team. 

Reviewer: Add slack link?

Author: That's in the footer of every page of documentation (which you don't see here, but will see when the doc is finally rendered on RTD).  If it's stated here, as well, it sounds clumsy, as it'll be placed above the exact same boilerplate.  

Of course, that *isn't* the case on github: you won't see the footer that RTD produces there.

@svekars, how should we handle this kind of thing?

> please engage with your Pachyderm support team. 

Reviewer 2: ```suggestion coordinate the deployment with your Pachyderm support team. 
```

### Role or ClusterRole

Depending on whether you used the `--local-roles` flag or not, the next manifest `kind` is `Role` or `ClusterRole`.

### RoleBinding or ClusterRoleBinding

This manifest binds the `Role` or `ClusterRole` to the `ServiceAccount` created above.

## Application-related

This section walks you through the application-related manifests.

### PersistentVolume

If you used `--static-etc-volume` to deploy Pachyderm,
the value that you specify for `--persistent-disk` causes `pachctl` to write a manifest for creating a [`PersistentVolume`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) that Pachyderm's `etcd` uses in its [`PersistentVolumeClaim`](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims).

A common persistent volume that is used in enterprises is an NFS mount backed by a storage fabric.
In this case, a StorageClass for an NFS mount is made available for consumption.
Consult with your Kubernetes administrators to learn what resources are available for your deployment.

### PersistentVolumeClaim

> If you deployed Pachyderm by using `--static-etc-volume`, the Pachyderm's `etcd` uses this `PersistentVolumeClaim`.

Reviewer: ```suggestion
If you deployed Pachyderm by using `--static-etcd-volume`, Pachyderm's `etcd` uses this `PersistentVolumeClaim`.
```

> See this manifest's name in the `etcd` Deployment manifest below.

Reviewer: Should this be, "you will see this manifest's name in the `etcd` Deployment manifest below." ? Or are you iterally telling the user to go to a different section?

### StorageClass

If you used the `--dynamic-etcd-nodes` flag to deploy Pachyderm,
this manifest specifies the kind of storage and provisioner that is appropriate for what you have specified in the `--persistent-disk` flag. 

> **Note:** You will not see this manifest if you specified `azure` as the argument to `--persistent-disk`.

Reviewer: Why? Add "because azure does blah blah instead..."

### Service

In a typical Pachyderm deployment, 
you see three [`Service`](https://kubernetes.io/docs/concepts/services-networking/service/) manifests. 
Services are how Kubernetes exposes Pods to the network.

> If you use StatefulSets to deploy Pachyderm,

Reviewer: > To make your `containerPort` values work properly, add environment variables to a `Deployment` or `StatefulSet` object.
You can verify this functionality in the [OpenShift](./openshift.html) example.	You can verify this functionality in the [OpenShift](./openshift.html) example.

Does one _need_ to add environment variables to make this work properly? What does "work properly" mean? Is this only required for openshift or always? Make this sentence clearer. If it's a require set of steps, it's pretty darn hidden away in this document. If it's "you can customize you deployment with these envvars" then thats fine, but still should clarify what these do and when they should be used.

### Editing your manifest to customize it further

This functionality requires an experienced Kubernetes administrator.
If you are attempting a highly customized deployment, 
use one of the Pachyderm support resources below.

### Deploying

> The command you'll want to run depends on the command you ran, above.

Reviewer: > The command you'll want to run depends on the command you ran, above. 

Does it? It's the same command, you just pass it the actual file you made. This section is literally just `kubectl apply -f ./whatever_manifest_you_created.json`. Not sure why there are 3 subsections
